src/devices/colordistancesensor.ts

```diff
@@ -79,11 +81,128 @@ export class ColorDistanceSensor extends Device {
+        let address = 0;
+        if (channel >= 4) {
+            channel -= 4;
+            address = 1;
+        }
+        const message = Buffer.alloc(2);
+        // Send "Single output mode"
+        message[0] = ((channel - 1) << 4) + (address << 3);
+        message[1] = 6;
+        return this.sendPFIRMessage(message);
+    }
+
+
+    /**
+     * Set the power of a Power Functions motor via IR
+     * @method ColorDistanceSensor#setPFPower
+     * @param {number} channel Channel number, between 1-4
+     * @param {string} output Outport port, "RED" (A) or "BLUE" (B)
+     * @param {number} power -7 (full reverse) to 7 (full forward). 0 is stop. 8 is brake.
+     * @returns {Promise} Resolved upon successful issuance of the command.
+     */
+    public setPFPower (channel: number, output: string, power: number) {
+        let address = 0;
+        if (channel > 4) {
+            channel -= 4;
+            address = 1;
+        }
+        const message = Buffer.alloc(2);
+        // Send "Single output mode"
+        message[0] = ((channel - 1) << 4) + (address << 3) + (output === "RED" ? 4 : 5);
+        message[1] = this._pfPowerToPWM(power) << 4;
+        return this.sendPFIRMessage(message);
+    }
+
+
+    /**
+     * Start Power Functions motors running via IR
+     *
+     * NOTE: This command is designed for bang-bang style operation. To keep the motors running, the sensor needs to be within range of the IR receiver constantly.
+     * @method ColorDistanceSensor#startPFMotors
+     * @param {Buffer} channel Channel number, between 1-4
+     * @param {Buffer} powerA -7 (full reverse) to 7 (full forward). 0 is stop. 8 is brake.
+     * @param {Buffer} powerB -7 (full reverse) to 7 (full forward). 0 is stop. 8 is brake.
+     * @returns {Promise} Resolved upon successful issuance of the command.
+     */
+    public startPFMotors (channel: number, powerA: number, powerB: number) {
+        let address = 0;
+        if (channel > 4) {
+            channel -= 4;
+            address = 1;
+        }
+        const message = Buffer.alloc(2);
+        // Send "Combo PWD mode"
+        message[0] = (((channel - 1) + 4 + (address << 3)) << 4) + this._pfPowerToPWM(powerA);
+        message[1] += this._pfPowerToPWM(powerB) << 4;
+        return this.sendPFIRMessage(message);
+    }
+
+
+    /**
+     * Send a raw Power Functions IR command
+     * @method ColorDistanceSensor#sendPFIRMessage
+     * @param {Buffer} message 2 byte payload making up a Power Functions protocol command. NOTE: Only specify nibbles 1-3, nibble 4 should be zeroed.
+     * @returns {Promise} Resolved upon successful issuance of the command.
+     */
+    public sendPFIRMessage (message: Buffer) {
+        if (this.isWeDo2SmartHub) {
+            throw new Error("Power Functions IR is not available on the WeDo 2.0 Smart Hub");
+        } else {
+            const payload = Buffer.alloc(2);
+            payload[0] = (message[0] << 4) + (message[1] >> 4);
```

Comment:
Why not generate the `message` matching the `payload` format to avoid this nibble shuffling? It seems that this is correct as far as I can tell, but it was hard for me to follow.

Reply:
Yeah, I agree it makes the code a little unclear, but my reasoning for this is to allow raw messages to be passed in in the exact order as specified in the LPF1.0 documentation. Since the hub accepts data in little endian format, the byte order doesn't match what the documentation shows, which could cause confusion to a user of this library.

It also made it easier for me to author the functions - since I could write the code to output data in the order the documentation specified, and not what the LPF2.0 hub expects.

The downside is a little bit of magical data shifting, but I think the clarity and user friendliness it provides is worth the tradeoff?

Reply:
OK, I see your point. I did not spot that this shuffling is equivalent to shift right by one nibble to account for missing checksum followed by an endianess swap. Though while I agree that big endian makes more sense in the context of a communication protocol, note that the LPF spec itself never needs to serialize the code into bytes. That is why in my solution I went for returning a 12 bit integer instead (https://gist.github.com/nutki/e5c2987ffb897a25ef3dce21ec27b8b5)